File: src/input_parsers/PropertyParser.cpp
/*********************                                                        */
/*! \file PropertyParser.cpp
 ** \verbatim
 ** Top contributors (to current version):
 **   Guy Katz
 ** This file is part of the Marabou project.
 ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
 ** in the top-level source directory) and their institutional affiliations.
 ** All rights reserved. See the file COPYING in the top-level source
 ** directory for licensing information.\endverbatim
 **
 ** [[ Add lengthier description here ]]

**/

#include "Debug.h"
#include "File.h"
#include "InputParserError.h"
#include "MStringf.h"
#include "PropertyParser.h"
#include "Pair.h"


enum VariableType {
    Input = 0,
    Output = 1,
    Hidden = 2
};

static Equation::EquationType parseEquationType( const String &token )
{
    if ( token == ">=" )
        return Equation::GE;
    if ( token == "<=" )
        return Equation::LE;
    if ( token == "=" )
        return Equation::EQ;

    throw InputParserError( InputParserError::UNEXPECTED_INPUT, token.ascii() );
}

static double parseScalar( const String &token )
{
    std::string::size_type end;
    double value = std::stod( token.ascii(), &end );
    if ( end != token.length() )
    {
        String errorMessage = Stringf( "%s not a scalar", token.ascii() );
        throw InputParserError( InputParserError::UNEXPECTED_INPUT, errorMessage.ascii() );
    }
    return value;
}

static double parseCoefficient ( const String &token )
{
    if ( token == "+" )
        return 1;
    else if ( token == "-" )
        return -1;
    else
        return atof( token.ascii() );
}

void PropertyParser::parse( const String &propertyFilePath, InputQuery &inputQuery )
{
    if ( !File::exists( propertyFilePath ) )
    {
        printf( "Error: the specified property file (%s) doesn't exist!\n", propertyFilePath.ascii() );
        throw InputParserError( InputParserError::FILE_DOESNT_EXIST, propertyFilePath.ascii() );
    }

    File propertyFile( propertyFilePath );
    propertyFile.open( File::MODE_READ );

    try
    {
        while ( true )
        {
            String line = propertyFile.readLine().trim();
            bool isEmpty = line.length() == 0;
            bool isComment = line.substring(0,2) == "//";
            if ( !isEmpty && !isComment )
            {
                parseEquation( line, inputQuery );
            }
        }
    }
    catch ( const CommonError &e )
    {
        // A "READ_FAILED" is how we know we're out of lines
        if ( e.getCode() != CommonError::READ_FAILED )
            throw e;
    }
}

void PropertyParser::parseEquation( const String &line, InputQuery &inputQuery )
{
    List<String> tokens = line.tokenize( " " );

    // Malformed, each line must have at least one variable, a relation and a coefficient.
    if ( tokens.size() < 3 ) {
        String errorMessage = Stringf( "Invalid equation '%s'", line.ascii() );
        throw InputParserError( InputParserError::UNEXPECTED_INPUT, errorMessage.ascii() );
    }

    auto it = tokens.rbegin();
    double scalar = parseScalar( *it );
    ++it;
    Equation::EquationType equationType = parseEquationType( *it );
    ++it;

    // Now extract the addends. In the special case where we only have
    // one addend, we add this equation as a bound. Otherwise, we add
    // as an equation.
    if ( tokens.size() == 3 )
    {
        // Special case: add as a bound on the variable
        String token = (*it).trim();
        Pair<double, unsigned> addend = parseAddend ( token, inputQuery );

        double coefficient = addend.first();
        // Currently don't support coefficients on bounds
        // (see https://github.com/NeuralNetworkVerification/Marabou/issues/625)
        if ( coefficient != 1 )
        {
            String errorMessage = Stringf( "'%s' - coefficients for bounds not supported", line.ascii() );
            throw InputParserError( InputParserError::UNEXPECTED_INPUT, errorMessage.ascii() );
        }

        unsigned variable = addend.second();

        if ( equationType == Equation::GE )
        {
            if ( inputQuery.getLowerBound( variable ) < scalar )
                inputQuery.setLowerBound( variable, scalar );
        }
        else if ( equationType == Equation::LE )
        {
            if ( inputQuery.getUpperBound( variable ) > scalar )
                inputQuery.setUpperBound( variable, scalar );
        }
        else
        {
            ASSERT( equationType == Equation::EQ );

            if ( inputQuery.getLowerBound( variable ) < scalar )
                inputQuery.setLowerBound( variable, scalar );
            if ( inputQuery.getUpperBound( variable ) > scalar )
                inputQuery.setUpperBound( variable, scalar );
        }
    }
    else
    {
        // Normal case: add as a new equation over multiple variables
        Equation equation( equationType );
        equation.setScalar( scalar );

        while ( it != tokens.rend() )
        {
            String token = (*it).trim();
            Pair<double, unsigned> addend = parseAddend ( token, inputQuery );
            double coefficient = addend.first();
            unsigned variable = addend.second();
            equation.addAddend( coefficient, variable );
            ++it;
        }

        inputQuery.addEquation( equation );
    }
}

// Returns (coefficient, variableNumber)
Pair<double, unsigned> PropertyParser::parseAddend ( const String &token , InputQuery &inputQuery )
{
    String symbol;
    VariableType variableType;
    unsigned variableTypeCount = 0;

    if ( token.contains ("x") )
    {
        variableType = VariableType::Input;
        symbol = "x";
        ++variableTypeCount;
    }

    if ( token.contains ("y") )
    {
        variableType = VariableType::Output;
        symbol = "y";
        ++variableTypeCount;
    }

    if ( token.contains("h") )
    {
        variableType = VariableType::Hidden;
        symbol = "h";
        ++variableTypeCount;
    }

    // Make sure that we have identified precisely one kind of variable
    if ( variableTypeCount != 1 )
    {
        String errorMessage = Stringf( "'%s' does not contain a valid variable name", token.ascii() );
        throw InputParserError( InputParserError::INVALID_VARIABLE_NAME, errorMessage.ascii() );
    }

    // Extract variable and coefficient strings
    List<String> subTokens = token.tokenize( symbol );
    String coefficientString;
    String variableIDString;
    if ( subTokens.size() == 1)
    {
        coefficientString = "+";
        variableIDString = *subTokens.begin();
    }
    else if ( subTokens.size() == 2 )
    {
        coefficientString = *subTokens.begin();
        variableIDString = *subTokens.rbegin();
    }
    else
    {
        String errorMessage = Stringf( "'%s' is not a valid addend", token.ascii() );
        throw InputParserError( InputParserError::UNEXPECTED_INPUT, errorMessage.ascii() );
    }

    // Parse coefficient
    double coefficient = parseCoefficient ( coefficientString );

    // Parse variable
    unsigned variable;
    if ( variableType == VariableType::Input )
    {
        unsigned justIndex = atoi( variableIDString.ascii() );
        ASSERT( justIndex < inputQuery.getNumOutputVariables() );
        variable = inputQuery.inputVariableByIndex( justIndex );
    }
    else if ( variableType == VariableType::Output )
    {
        unsigned justIndex = atoi( variableIDString.ascii() );
        ASSERT( justIndex < inputQuery.getNumOutputVariables() );
        variable = inputQuery.outputVariableByIndex( justIndex );
    }
    else
    {
        // These variables are of the form h_2_5
        subTokens = variableIDString.tokenize( "_" );

        if ( subTokens.size() != 2 )
        {
            String errorMessage = Stringf( "'%s' does not contain valid hidden variable indices", token.ascii() );
            throw InputParserError( InputParserError::INVALID_HIDDEN_VARIABLE_INDICES, errorMessage.ascii() );
        }

        auto subToken = subTokens.begin();
        unsigned layerIndex = atoi( subToken->ascii() );
        ++subToken;
        unsigned nodeIndex = atoi( subToken->ascii() );

        NLR::NetworkLevelReasoner *nlr = inputQuery.getNetworkLevelReasoner();
        if ( !nlr )
        {
            String errorMessage = Stringf( "Cannot refer to hidden variable '%s' when network level reasoning is disabled", token.ascii() );
            throw InputParserError( InputParserError::NETWORK_LEVEL_REASONING_DISABLED, errorMessage.ascii() );
        }

        if ( nlr->getNumberOfLayers() < layerIndex )
        {
            String errorMessage = Stringf( "Layer index '%d' for hidden variable '%s' is invalid", layerIndex, token.ascii() );
            throw InputParserError( InputParserError::HIDDEN_VARIABLE_DOESNT_EXIST_IN_NLR );
        }

        const NLR::Layer *layer = nlr->getLayer( layerIndex );
        if ( layer->getSize() < nodeIndex || !layer->neuronHasVariable( nodeIndex ) )
        {
            String errorMessage = Stringf( "Node index '%d' for hidden variable '%s' is invalid", nodeIndex, token.ascii() );
            throw InputParserError( InputParserError::HIDDEN_VARIABLE_DOESNT_EXIST_IN_NLR );
        }

        variable = layer->neuronToVariable( nodeIndex );
    }

    return Pair<double,unsigned>(coefficient, variable);
}

//
// Local Variables:
// compile-command: "make -C ../.. "
// tags-file-name: "../../TAGS"
// c-basic-offset: 4
// End:
//

File: maraboupy/PropertyParser.py

'''
/* *******************                                                        */
/*! \file PropertyParser.py
 ** \verbatim
 ** Top contributors (to current version):
 ** Alex Usvyatsov
 ** This file is part of the Marabou project.
 ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
 ** in the top-level source directory) and their institutional affiliations.
 ** All rights reserved. See the file COPYING in the top-level source
 ** directory for licensing information.\endverbatim
 **
 ** \brief
 ** Class that parses a property file into a list of strings representing the individual properties.
 **
 ** [[ Add lengthier description here ]]
 **/
'''



import sys
import re

types_of_property_by_variables = ['x', 'y', 'ws', 'm']
'''
    'x'     : input property (mentions only input variables)
    'y'     : output property (mentions only output variables)
    'ws'    : hidden variable property (mentions only hidden variables)
    'm'     : mixed

'''


def parseProperty(property_filename):

    '''
        Parses a property file using regular expressions
        Replaces occurrences of x?? (where ?? are digits) by x[??]
        Replaces occurrences of y?? (where ?? are digits) by y[??]
        (for convenience of evaluating the expression with python's parser)
        Returns:
             two dictionaries: equations and bounds
                each dictionary has as keys the type (types_of_property_by_variables) of property:
                'x','y','m',(mixed),'ws'

                values are lists of properties of the appropriate type
                    e.g., bounds['x'] is a list of bounds on input variables, where x?? has ben replaced by x[??]
                    so for example, bounds['x'] can look like this: ['x[0] >= 0', 'x[1] <= 0.01']

             a list of all properties, given as a list of tuples of strings:
             (class_of_property,type_of_property,property,index)
                where:
                    class_of_property is 'e' (equation) or 'b' (bound)
                    type_of_property is 'x','y','ws', or 'm' (for 'mixed'),
                    property is a line from the property file (unchanged)
                    index is the index of a bound/equation in the appropriate list

    '''

    properties = {'x': [], 'y': [], 'ws': [], 'm': []}

    equations = {'x': [], 'y': [], 'ws': [], 'm': []}
    bounds = {'x': [], 'y': [], 'ws': [], 'm': []}

    reg_input = re.compile(r'[x](\d+)')
    # matches a substring of the form x??? where ? are digits

    reg_output = re.compile(r'[y](\d+)')
    # matches a substring of the form y??? where ? are digits

    reg_ws = re.compile(r'[w][s][_](\d+)[_](\d+)')
    # matches a substring of the form ws_???_??? where ? are digits

    reg_equation = re.compile(r'[+-][xy](\d+) ([+-][xy](\d+) )+(<=|>=|=) [+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?$')
    # matches a string that is a legal equation with input (x??) or output (y??) variables


    reg_bound = re.compile(r'[xy](\d+) (<=|>=|=) [+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?')
    # matches a string which represents a legal bound on an input or an output variable

    reg_ws_bound = re.compile(r'[w][s][_](\d+)[_](\d+) (<=|>=|=) [+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?')
    # matches a string which represents a legal bound on a hidden variable

    try:
        with open(property_filename) as f:
            line = f.readline()
            while(line):

                type_of_property = ''

                # Computing type_of_property: input/output/ws/mixed
                if (reg_input.search(line)): # input variables present
                    type_of_property = 'x'
                if (reg_output.search(line)): # output variables present
                    type_of_property = 'm' if type_of_property else 'y'
                if (reg_ws.search(line)): # hidden variable present
                    type_of_property = 'm' if type_of_property else 'ws'

                if not type_of_property:
                    print('An expression of an unknown type found while attempting to parse '
                          'property file ', property_filename)
                    print('expression: ', line)
                    sys.exit(1)

                if reg_equation.match(line):  # Equation

                    #replace xi by x[i] and yi by y[i]
                    new_str = line.strip()
                    new_str = reg_input.sub(r"x[\1]", new_str)

                    new_str = reg_output.sub(r"y[\1]", new_str)

                    # Debug
                    print('equation')
                    print(new_str)

                    index = len(equations[type_of_property])

                    equations[type_of_property].append(new_str)

                    class_of_property='e'

                elif reg_bound.match(line):  # I/O Bound

                    # replace xi by x[i] and yi by y[i]
                    new_str = line.strip()
                    new_str = reg_input.sub(r"x[\1]", new_str)

                    new_str = reg_output.sub(r"y[\1]", new_str)

                    print('bound: ', new_str)  #Debug

                    index = len(bounds[type_of_property])

                    bounds[type_of_property].append(new_str)

                    class_of_property = 'b'

                else:
                    assert reg_ws_bound.match(line)  # At this point the line has to match a legal ws bound

                    index = len(bounds['ws'])
                    bounds['ws'].append(line.strip()) # Storing without change

                    class_of_property = 'b'  # Perhaps at some point better add a new type for ws_bound?

                properties[type_of_property].append({'class_of_property': class_of_property,
                                                     'type_of_property': type_of_property,'line': line, 'index': index})

                line = f.readline()
        print('successfully read property file: ', property_filename)
        return equations, bounds, properties

    except:
        print('Something went wrong while parsing the property file', property_filename)
        sys.exit(1)

